### PR TITLE
Joe/checkpointing

### PR DESCRIPTION
**Pull request by visionjo**

### CODE REVIEW

File: docs/source/best-practices/checkpointing.rst [hunk 1]
```diff
@@ -2,21 +2,21 @@
 Checkpoint/Restart Discovery Jobs
 *****************************************
 
-The complexity of HPC systems may introduce unpredictable behaviors and may result in job failures due to hardware or software. Applying fault tolerance techniques to your HPC workflows allows your jobs to become more resilient to crashes, partition time limits and hardware failures.
+The complexity of HPC systems may introduce unpredictable behaviors and result in job failures due to hardware or software. Applying fault tolerance techniques to your HPC workflows makes your jobs more resilient to crashes, partition time limits, and hardware failures.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:5]: Since we don't discuss any other "fault tolerance techniques" in this page, can we remove, "Applying fault tolerance techniques to your HPC workflows . . ."  And maybe just take the checkpointing definition from the next subsection and move that up to this paragraph?

File: docs/source/best-practices/checkpointing.rst [hunk 2]
```diff
@@ -2,21 +2,21 @@
...
 
  * Create resilient workflows in the existence of faults.
  * Overcome most scheduler resource time limitations.
- * Implement an early error detection approach by inspecting intermediate results.  
+ * Implement an early error detection approach by inspecting intermediate results.
 
 The Checkpointing technique
 ================================
 
 Checkpointing is a fault tolerance technique designed to overcome the “fail-stop” failure type (interruption of the execution of a job). It is based on the BER technique (Backward error recovery or Rollback-recovery algorithm):
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:16]: For clarity, can we change to something like: "Checkpointing is a fault tolerance technique based on the backward error recovery (BER) technique and is designed to overcome the “fail-stop” failure type (interruption of the execution of a job)."

File: docs/source/best-practices/checkpointing.rst [hunk 3]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
 Checkpointing types
 ================================
 
-Checkpointing can be implemented in different levels of your workflow:
+Implement checkpoints at different levels of your workflow:
 
-  * User-level checkpointing - suitable if you develop your own code, or have sufficient knowledge of the application code to integrate checkpointing techinques yourself. Generally, this approach is not recommended for most Discovery users.
-  * Application-level checkpointing - recommended for most Discovery users. Utilize the checkpointing tool that is already available in your software application. Most software designed for HPC have a checkpointing option, and information on proper usage is often available in the software user manual. 
-  * System-level checkpointing - done on the system side, where the state of the entire process is being saved. This option is less efficient than User-level or Application-level checkpointing as it introduces a lot of redundancy.   
+  * User-level checkpointing - suitable if you develop your code or have sufficient knowledge of the application code to integrate checkpointing techniques. We recommend this approach for some Discovery users.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:30]: I think we should define "some Discovery users." Either include which Discovery users (advanced, proficient with checkpointing, or whatever) or change back to something like, this method is not typically recommended.

File: docs/source/best-practices/checkpointing.rst [hunk 4]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+  * Application-level checkpointing - recommended for most Discovery users. Utilize the checkpointing tool that is already available in your software application. For example, most software designed for HPC has a checkpointing option, and information on proper usage is often available in the software user manual.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:31]: I would suggest moving this bullet to the top if it's the recommended method for checkpointing.

File: docs/source/best-practices/checkpointing.rst [hunk 5]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+  * System-level checkpointing - done on the system side, where we save the state of the entire process. This option is less efficient than User-level or Application-level checkpointing as it introduces a lot of redundancy.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:32]: Remove the "we," I'm not sure if this means that the RC team saves the process, but if that is the procedure, maybe change to, "the systems team saves the state of the entire process." Or whatever makes the most sense.

File: docs/source/best-practices/checkpointing.rst [hunk 6]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+  * Model-level checkpointing - training models are often the intent of the user doing machine learning on Discovery.
+
+Which checkpointing method to use?
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:35]: I like this addition, but would it make more sense to maybe create a table/graphic and combine it with the last section? So we can define the checkpointing type/method, and then add the recommended method? 

Also, we should probably be consistent with the terminology and stick to either type or method just for clarity.

File: docs/source/best-practices/checkpointing.rst [hunk 7]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+  Some packages and functions allow for easy checkpointing if you are developing code using Python, Matlab, or R. Some examples include `Python PyTorch checkpointing <https://pytorch.org/tutorials/recipes/recipes/saving_and_loading_a_general_checkpoint.html>`_, `TensorFlow checkpointing <https://www.tensorflow.org/guide/checkpoint>`_, `Python Pickle checkpointing <https://deap.readthedocs.io/en/master/tutorials/advanced/checkpoint.html>`_, `MATLAB checkpointing <https://www.mathworks.com/help/gads/work-with-checkpoint-files.html>`_ and `R checkpointing <https://cran.r-project.org/web/packages/checkpoint/vignettes/checkpoint.html>`_. Additionally, many Computational Chemistry and Molecular Dynamics software have built-in checkpointing options, such as `GROMACS <https://manual.gromacs.org/documentation/current/user-guide/managing-simulations.html>`_ and `LAMMPS <https://docs.lammps.org/restart.html>`_.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:44]: For clarity, I would recommend the following change: Python PyTorch checkpointing, TensorFlow checkpointing, Python Pickle checkpointing, MATLAB checkpointing and R checkpointing allow for easy checkpointing if you are developing code using Python, Matlab, or R. Additionally, many Computational Chemistry and Molecular Dynamics software have built-in checkpointing options, such as GROMACS and LAMMPS.

File: docs/source/best-practices/checkpointing.rst [hunk 8]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
 
 
-Implementing checkpointing can be acheived by:
- * Some save-and-load mechanism of your calculation state. 
- * The use of `Slurm Job Arrays <https://slurm.schedmd.com/job_array.html>`_. 
+Implementing checkpointing can be achieved by:
+ * Some save-and-load mechanisms of your calculation state.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:48]: Can this be more fully explained? Or, can we remove "some?"

File: docs/source/best-practices/checkpointing.rst [hunk 9]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+  Some packages and functions allow for easy checkpointing if you are developing code using Python, Matlab, or R. Some examples include `Python PyTorch checkpointing <https://pytorch.org/tutorials/recipes/recipes/saving_and_loading_a_general_checkpoint.html>`_, `TensorFlow checkpointing <https://www.tensorflow.org/guide/checkpoint>`_, `Python Pickle checkpointing <https://deap.readthedocs.io/en/master/tutorials/advanced/checkpoint.html>`_, `MATLAB checkpointing <https://www.mathworks.com/help/gads/work-with-checkpoint-files.html>`_ and `R checkpointing <https://cran.r-project.org/web/packages/checkpoint/vignettes/checkpoint.html>`_. Additionally, many Computational Chemistry and Molecular Dynamics software have built-in checkpointing options, such as `GROMACS <https://manual.gromacs.org/documentation/current/user-guide/managing-simulations.html>`_ and `LAMMPS <https://docs.lammps.org/restart.html>`_.
 
 
-Implementing checkpointing can be acheived by:
- * Some save-and-load mechanism of your calculation state. 
- * The use of `Slurm Job Arrays <https://slurm.schedmd.com/job_array.html>`_. 
+Implementing checkpointing can be achieved by:
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:47]: Can we change this to "Implement checkpointing:" or "How to Implement checkpointing?"

File: docs/source/best-practices/checkpointing.rst [hunk 10]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+ * Some save-and-load mechanisms of your calculation state.
+ * The use of `Slurm Job Arrays <https://slurm.schedmd.com/job_array.html>`_.
 
 .. note::
-   To overcome partition time limits, replace your single long job with multiple shorter jobs. Using job arrays, set each job to run one after the other. Each job will write a checkpoint file if checkpointing is implemented. The next job in line will be the latest checkpoint file to continue from the latest state of the calculation.
+   To overcome partition time limits, replace your single long job with multiple shorter jobs. Then, using job arrays, set each job to run one after the other. Each job will write a checkpoint file if implemented. For example, the following job in line will be the latest checkpoint file to continue from the latest state of the calculation.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:52]: From an outside view (having no idea how to actually checkpoint or what job arrays are), the previous version of this paragraph seems easier to follow.

File: docs/source/best-practices/checkpointing.rst [hunk 11]
```diff
@@ -25,25 +25,37 @@ Checkpointing is a fault tolerance technique designed to overcome the “fail-st
...
+
+Application-level checkpointing
+===============================
 
 GROMACS checkpointing example
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+-----------------------------
 
 This example demonstrates how to implement a longer `GROMACS <https://www.gromacs.org/>`_ job of 120 hours by using multiple shorter jobs on the **short** partition. We use Slurm job arrays and the GROMACS built-in checkpointing option (read more `here <https://manual.gromacs.org/documentation/current/user-guide/managing-simulations.html>`_) to implement checkpointing.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:60]: Would it make sense to change the first sentence to: The following example demonstrates how to use Slurm job arrays to break a 120-hour GROMACS job into multiple shorter jobs on the short partition.

File: docs/source/best-practices/checkpointing.rst [hunk 12]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
 
    sbatch submit_mdrun_array.bash
 
-Python TensorFlow checkpointing example
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Checkpointing using DMTCP
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:89]: For consistency, this should either be DMTCP checkpointing example, or the previous heading should be Checkpointing using GROMACS.

File: docs/source/best-practices/checkpointing.rst [hunk 13]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+--------------------------
+
+`DMTCP <https://dmtcp.sourceforge.io/>`_ (Distributed MultiThreaded checkpointing) is a tool that lets you checkpoint without the need to change your code. It Works with most Linux applications such as Python, Matlab, R, GUI, MPI, etc.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:92]: [DMTCP](https://dmtcp.sourceforge.io/) (Distributed MultiThreaded checkpointing) is available on the cluster, allows checkpointing without the need to change your code, and works with most Linux applications (e.g., Python, Matlab, R, GUI, and MPI). DMTCP runs in the background of your program without causing significant performance loss and saves the process states into checkpoint files.

File: docs/source/best-practices/checkpointing.rst [hunk 14]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+The program runs in the background of your program without significant performance loss and saves the process states into checkpoint files. DMTCP is available on the cluster ::
+
+ module avail dmtcp
+ module show dmtcp
+ module load dmtcp/2.6.0
+
+As DMTCP runs in the background, it requires some changes to your Shell script. For examples of how to checkpoint with DMTCP visit `here <https://github.com/northeastern-rc/training-checkpointing/tree/main/Exercise_3>`_.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:99]: Because DMTCP runs in the background, changes to your shell script are required.

File: docs/source/best-practices/checkpointing.rst [hunk 15]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+The example demonstrates how to use DMTCP with a simple C++ program (scripts modified from `RSE-Cambridge <https://github.com/RSE-Cambridge/dmtcp-tests>`_).
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:100]: In the previous sentence, we refer to 'examples' but in this sentence just one example, so one should be updated to align with the other. Also, there are 3 exercises shown when the link is clicked, so it would be good to direct to the specific exercise that aligns with this section (if necessary).

File: docs/source/best-practices/checkpointing.rst [hunk 16]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+
+Tips and Tricks
+---------------------
+
+What data to save?
+ * Non-temporary application data
+ * Any application data that has changed since the last checkpoint
+ * Delete no longer proper checkpoints - keep only the most recent checkpoint file.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:108]: Change to: Keep only the most recent checkpoint file and make sure to remove the period.

File: docs/source/best-practices/checkpointing.rst [hunk 17]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+
+How frequently should we checkpoint?
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:110]: Take out bullet points and make this section something like: Consider how long it takes to a checkpoint and restart your calculation, in most cases, a checkpointing rate of every 10-15 minutes is sufficient. Keep in mind, if you checkpoint too often, you will slow down your calculation, but if you checkpoint too infrequently, you risk large/long rollback times.

File: docs/source/best-practices/checkpointing.rst [hunk 18]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+ * Too infrequently – leads to large/long rollback times.
+ * Consider how long it takes to a checkpoint and restart your calculation.
+ * In most cases, a rate of every 10-15 minutes is ok.
+
+.. _ML Model-level:
+
+ML Model-level checkpointing
+============================
+
+Checkpointing is a technique used in training machine learning models to save the model's state during the training process periodically. Doing this allows the training process to be resumed from the saved checkpoint if the training is interrupted or terminated so that the progress made so far is not lost. The saved state typically includes the model's parameters, optimizer state, and training information, such as the epoch number and loss value (or the accuracy).
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:121]: Consider being more specific with the term being defined, Model - level checkpointing, rather than just checkpointing to be more clear for readers that might be jumping sections. I know it's in the heading, but it's better to be as specific as possible.

Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:121]: Model-level checkpointing is a technique used to periodically save the machine learning (ML) model’s state while training ML models. This checkpointing allows the training process to be resumed from the saved checkpoint if the training is interrupted or terminated. The saved state typically includes the model’s parameters, optimizer state, and training information such as the epoch number and loss value (or the accuracy).

File: docs/source/best-practices/checkpointing.rst [hunk 19]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+
+The following instructions are helpful for long-running training jobs, as they enable faster recovery from failures and better tracking of the training process.
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:123]: No need to add this as a new paragraph, can be added to the previous paragraph.

File: docs/source/best-practices/checkpointing.rst [hunk 20]
```diff
@@ -74,8 +86,53 @@ To submit this array job to the scheduler, use the following command::
...
+
+Why Checkpointing is Important in Deep Learning?
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:125]: To follow the style of the other paragraphs, only capitalize the first word and proper nouns.

File: docs/source/best-practices/checkpointing.rst [hunk 21]
```diff
@@ -110,45 +167,67 @@ Where the checkpointing implementation is given in this code snippet of ``train_
...
+Save the checkpoint file to multiple locations, such as a local drive and the cloud, to ensure that
+the checkpoint is recovered in case of failure.
+
+Use the latest versions of libraries
+""""""""""""""""""""""""""""""""""""""""
+Using the latest version of PyTorch and other relevant libraries is vital; changes in these libraries may cause
+compatibility issues with older checkpoints. With these best practices, you can ensure that your PyTorch models are
+saved efficiently and effectively and that your progress is not lost in case of a crash or interruption.
+
+Avoid saving unnecessary information
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:206]: Can the information from this section be combined with the Save only the model's Sate_dict section?

File: docs/source/best-practices/checkpointing.rst [hunk 22]
```diff
@@ -110,45 +167,67 @@ Where the checkpointing implementation is given in this code snippet of ``train_
...
- * Any application data that has been modified since the last checkpoint
- * Delete checkpoints that are no longer useful - keep only the most recent checkpoint file.
+  sbatch submit_tf_array.bash
 
-How frequently to checkpoint? 
- * Too often – will slow down your calculation, may be I/O heavy and memory-limited.
- * Too infrequently – leads to large/long rollback times.
- * Consider how long it takes to checkpoint and restart your calculation. 
- * In most cases a rate of every 10-15 minutes is ok.
+Python PyTorch
```
Comment (tketchem14) [inline docs/source/best-practices/checkpointing.rst:179]: Is this heading a placeholder or can we change to Python Pytorch tips and tricks?

**Comment by visionjo** [comment]
Oh, this one was not ready yet. Unsure how you got assigned. My bad. I will circle back when i return to this soon! Thanks @tketchem14
